test/common/http/http1/codec_impl_test.cc

```diff
@@ -69,6 +115,76 @@ TEST_F(Http1ServerConnectionImplTest, Http10) {
 }
 
+TEST_F(Http1ServerConnectionImplTest, Http10AbsoluteNoOp) {
+  TestHeaderMapImpl expected_headers{{":path", "/"}, {":method", "GET"}};
+  Buffer::OwnedImpl buffer("GET / HTTP/1.0\r\n\r\n");
+  ExpectHeadersTest(Protocol::Http10, true, buffer, expected_headers);
+}
+
+TEST_F(Http1ServerConnectionImplTest, Http10Absolute) {
+  TestHeaderMapImpl expected_headers{
+      {":authority", "www.somewhere.com"}, {":path", "/foobar"}, {":method", "GET"}};
+  Buffer::OwnedImpl buffer("GET http://www.somewhere.com/foobar HTTP/1.0\r\n\r\n");
+  ExpectHeadersTest(Protocol::Http10, true, buffer, expected_headers);
+}
+
+TEST_F(Http1ServerConnectionImplTest, Http11AbsolutePath1) {
+  TestHeaderMapImpl expected_headers{
+      {":authority", "www.somewhere.com"}, {":path", "/"}, {":method", "GET"}};
+  Buffer::OwnedImpl buffer("GET http://www.somewhere.com/ HTTP/1.1\r\nHost: bah\r\n\r\n");
+  ExpectHeadersTest(Protocol::Http11, true, buffer, expected_headers);
+}
+
+TEST_F(Http1ServerConnectionImplTest, Http11AbsolutePath2) {
+  TestHeaderMapImpl expected_headers{
+      {":authority", "www.somewhere.com"}, {":path", "/"}, {":method", "GET"}};
+  Buffer::OwnedImpl buffer("GET http://www.somewhere.com/ HTTP/1.1\r\nHost: bah\r\n\r\n");
+  ExpectHeadersTest(Protocol::Http11, true, buffer, expected_headers);
+}
+
+TEST_F(Http1ServerConnectionImplTest, Http11AbsoluteEnabledNoOp) {
+  TestHeaderMapImpl expected_headers{
+      {":authority", "bah"}, {":path", "/foo/bar"}, {":method", "GET"}};
+  Buffer::OwnedImpl buffer("GET /foo/bar HTTP/1.1\r\nHost: bah\r\n\r\n");
+  ExpectHeadersTest(Protocol::Http11, true, buffer, expected_headers);
+}
+
+TEST_F(Http1ServerConnectionImplTest, Http11InvalidRequest) {
+  Buffer::OwnedImpl buffer("GET www.somewhere.com HTTP/1.1\r\nHost: bah\r\n\r\n");
+  Expect400(Protocol::Http11, true, buffer);
+}
+
+TEST_F(Http1ServerConnectionImplTest, Http11AbsolutePathNoSlash) {
+  TestHeaderMapImpl expected_headers{
+      {":authority", "www.somewhere.com"}, {":path", "/"}, {":method", "GET"}};
+  Buffer::OwnedImpl buffer("GET http://www.somewhere.com HTTP/1.1\r\nHost: bah\r\n\r\n");
+  ExpectHeadersTest(Protocol::Http11, true, buffer, expected_headers);
+}
+
+TEST_F(Http1ServerConnectionImplTest, Http11AbsolutePathBad) {
+  Buffer::OwnedImpl buffer("GET * HTTP/1.1\r\nHost: bah\r\n\r\n");
+  Expect400(Protocol::Http11, true, buffer);
+}
+
+TEST_F(Http1ServerConnectionImplTest, Http11AbsolutePortTooLarge) {
+  Buffer::OwnedImpl buffer("GET http://foobar.com:1000000 HTTP/1.1\r\nHost: bah\r\n\r\n");
+  Expect400(Protocol::Http11, true, buffer);
+}
+
+TEST_F(Http1ServerConnectionImplTest, Http11RelativeOnly) {
+  TestHeaderMapImpl expected_headers{
+      {":authority", "bah"}, {":path", "http://www.somewhere.com/"}, {":method", "GET"}};
+  Buffer::OwnedImpl buffer("GET http://www.somewhere.com/ HTTP/1.1\r\nHost: bah\r\n\r\n");
+  ExpectHeadersTest(Protocol::Http11, false, buffer, expected_headers);
+}
```

Comment:
In my time working on Google's reverse proxy we've seen a lot of "unexpected" fully qualified URLs (probably some assembled by corporate proxies and others by malware) which when passed on were not historically handled correctly by the next hop.  So I'm a fan of absolute url handling by default since unexpected things happen on the internet.

That said to @mattwoodyard 's point I definitely hadn't thought out the security aspects of it.  I'd be inclined to always support fully qualified URLs but 400 if the host header doesn't match the host in the URL field on the assumption if you're sending mismatched headers you're doing something shady, but I don't know how @mattklein123 feels about Envoy blocking shady-but-not-illegal traffic.

I'd say leave it in for now, and I'll poke around with our security folk to see if there's a good reason to default it.  I'll be reusing the HTTP1Settings for flow control so if we decide to kill off this config boolean later the structures around it will still be used later.